tests/v3/test_sync.py

```diff
@@ -0,0 +1,118 @@
+from collections.abc import AsyncGenerator
+import asyncio
+from unittest.mock import patch
+
+from zarr.v3.sync import sync, _get_loop, _get_lock, SyncError, SyncMixin
+from zarr.v3.config import SyncConfiguration
+
+import pytest
+
+
+@pytest.fixture(params=[True, False])
+def sync_loop(request) -> asyncio.AbstractEventLoop | None:
+    if request.param is True:
+        return _get_loop()
+
+    if request.param is False:
+        return None
+
+
+def test_get_loop() -> None:
+    # test that calling _get_loop() twice returns the same loop
+    loop = _get_loop()
+    loop2 = _get_loop()
+    assert loop is loop2
+
+
+def test_get_lock() -> None:
+    # test that calling _get_lock() twice returns the same lock
+    lock = _get_lock()
+    lock2 = _get_lock()
+    assert lock is lock2
+
+
+def test_sync(sync_loop: asyncio.AbstractEventLoop | None) -> None:
+    async def foo() -> str:
+        return "foo"
+
+    assert sync(foo(), loop=sync_loop) == "foo"
+
+
+def test_sync_raises(sync_loop: asyncio.AbstractEventLoop | None) -> None:
+    async def foo() -> str:
+        raise ValueError("foo")
+
+    with pytest.raises(ValueError):
+        sync(foo(), loop=sync_loop)
+
+
+def test_sync_raises_if_no_coroutine(sync_loop: asyncio.AbstractEventLoop | None) -> None:
+    def foo() -> str:
+        return "foo"
+
+    with pytest.raises(TypeError):
+        sync(foo(), loop=sync_loop)
+
+
+@pytest.mark.filterwarnings("ignore:coroutine.*was never awaited")
+def test_sync_raises_if_loop_is_closed() -> None:
+    loop = _get_loop()
+
+    async def foo() -> str:
+        return "foo"
+
+    with patch.object(loop, "is_closed", return_value=True):
+        with pytest.raises(RuntimeError):
+            sync(foo(), loop=loop)
+
+
+@pytest.mark.filterwarnings("ignore:coroutine.*was never awaited")
```

Comment:
despite this, we're still seeing the below warning. Why?

```
sys:1: RuntimeWarning: coroutine 'test_sync_raises_if_calling_sync_from_within_a_running_loop.<locals>.foo' was never awaited
RuntimeWarning: Enable tracemalloc to get the object allocation traceback
```

Reply:
This warning comes after the test has finished, during garbage collection. Maybe doing `del` on the objects (and closure) would fix. Or run the coroutine after the check.